pkg/services/ngalert/notifier/alertmanager_test.go

```diff
@@ -346,3 +348,80 @@ func TestPutAlert(t *testing.T) {
 		})
 	}
 }
+
+// Tests cleanup of expired Silences. We rely on prometheus/alertmanager for
+// our alert silencing functionality, so we rely on its tests. However, we
+// implement a custom maintenance function for silences, because we snapshot
+// our data differently, so we test that functionality.
+func TestSilenceCleanup(t *testing.T) {
+	require := require.New(t)
+	assert := assert.New(t)
+
+	// TODO: We have to use a real clock because Prometheus doesn't support clock
+	// injection. So modify the retention constant. This isn't great - we should
+	// make this an argument to the alertmanager so we can safely change it.
+	oldRetention := retentionNotificationsAndSilences
+	retentionNotificationsAndSilences = 30 * time.Millisecond
+	oldMaintenance := silenceMaintenanceInterval
+	silenceMaintenanceInterval = 15 * time.Millisecond
+	defer func() {
+		retentionNotificationsAndSilences = oldRetention
+		silenceMaintenanceInterval = oldMaintenance
+	}()
+
+	am := setupAMTest(t)
+	now := time.Now()
+	dt := func(t time.Time) strfmt.DateTime { return strfmt.DateTime(t) }
+
+	makeSilence := func(comment string, createdBy string,
+		startsAt, endsAt strfmt.DateTime, matchers amv2.Matchers) *apimodels.PostableSilence {
+		return &apimodels.PostableSilence{
+			ID: "",
+			Silence: amv2.Silence{
+				Comment:   &comment,
+				CreatedBy: &createdBy,
+				StartsAt:  &startsAt,
+				EndsAt:    &endsAt,
+				Matchers:  matchers,
+			},
+		}
+	}
+
+	tru := true
+	testString := "testName"
+	matchers := amv2.Matchers{&amv2.Matcher{Name: &testString, IsEqual: &tru, IsRegex: &tru, Value: &testString}}
+	// Create silences - one in the future, one currently active, one expired but
+	// retained, one expired and not retained.
+	silences := []*apimodels.PostableSilence{
+		// Active in future
+		makeSilence("", "tests", dt(now.Add(5*time.Hour)), dt(now.Add(6*time.Hour)), matchers),
+		// Active now
+		makeSilence("", "tests", dt(now.Add(-5*time.Hour)), dt(now.Add(6*time.Hour)), matchers),
+		// Expiring soon
+		makeSilence("", "tests", dt(now.Add(-5*time.Hour)), dt(now.Add(2*time.Second)), matchers),
+		// Expiring *very* soon
+		makeSilence("", "tests", dt(now.Add(-5*time.Hour)), dt(now.Add(1*time.Millisecond)), matchers),
+	}
+
+	for _, s := range silences {
+		_, err := am.CreateSilence(s)
+		require.NoError(err)
+	}
+
+	// Let enough time pass for the maintenance window to run.
+	tick := time.NewTicker(1500 * time.Millisecond)
+	<-tick.C
+
+	// So, what silences do we have now?
+	found, err := am.ListSilences(nil)
+	require.NoError(err)
+	assert.Len(found, 3)
+
+	// Wait again for the expired silence to fall out of retention.
+	<-tick.C
+
+	// So, what silences do we have now?
+	found, err = am.ListSilences(nil)
+	require.NoError(err)
+	require.Len(found, 2)
```

Comment:
```suggestion
	// So, what silences do we have now?
	require.Eventually(func() bool {
		found, err := am.ListSilences(nil)
		return err == nil && len(found) == 3
	}, 1500*time.Millisecond, 500*time.Millisecond)

	// So, what silences do we have now?
	require.Eventually(func() bool {
		found, err := am.ListSilences(nil)
		return err == nil && len(found) == 2
	}, 4*time.Second, 500*time.Millisecond)
```

The use of `require.Eventually` is technically a nit, but if possible, I would avoid tests that never timeout (L421). If you decide to stick with your approach of using the ticker, then how do you feel about adding a timeout?

Reply:
I didn't know about require.Eventually. Should definitely have a timeout here.